### DISCUSSION TITLE
after images warp_perspective I got a image tensor in which all element is 0.

### DISCUSSION THREAD

**Discussion by raphaelhou25** [kornia/kornia#2420] [Questions]
https://kornia.readthedocs.io/en/latest/geometry.transform.html#kornia.geometry.transform.warp_perspective

my code shows at below:

images_warped = kgt.warp_perspective(images, hg_mtxs, configs.output_2d_shape, padding_mode="zeros")

the images[0] is:
![image](https://github.com/kornia/kornia/assets/37241024/02b209e7-01b5-4d95-8aac-50541abd78c5)

homography_matix = [0.3030404, -0.16531825, -0.36811006, 0.25540864, -0.43341982, -0.012766998, 0.33119756, 0.42360407, 1.0]

this code runs on cuda.

but after warp_perspective, the images_warped is

![image](https://github.com/kornia/kornia/assets/37241024/33c83488-f7f1-4421-8459-6165742012d3)

why the images_warped image tensor all element is 0.

**Comment by edgarriba** [comment]
Is you image in the range of 0/1 and your homography in pixel space ?

**Reply by raphaelhou25** [reply]
Thank you for your prompt response.

1 my image tensor is not in range [0,1]
![image](https://github.com/kornia/kornia/assets/37241024/31522db2-f81e-4980-a26e-8c07596405bb)

2 my homography matrix is producted by network, then i didn't normalize it.

![image](https://github.com/kornia/kornia/assets/37241024/c6a6d693-41ad-43c5-b455-5578dc18311c)

my network structure shows at below:
```
class HG_CNN(nn.Module):
    def __init__(self):
        super(HG_CNN, self).__init__()
        self.layer1 = nn.Sequential(
            nn.Conv2d(96, 64, kernel_size=3, padding=1, bias=False),
            nn.BatchNorm2d(64),
            nn.ReLU(),
            nn.Conv2d(64, 32, kernel_size=3, padding=1, bias=False),
            nn.BatchNorm2d(32),
            nn.ReLU()
        )
        self.fc = nn.Sequential(
            nn.Dropout(0.5),
            nn.Linear(32 * 18 * 32, 1024),
            # nn.BatchNorm1d(1024),
            nn.ReLU(),
            nn.Dropout(0.5),
            nn.Linear(1024, 8)
        )
        hg_mlp_init_module(self.layer1)

    def forward(self, x):
        out = self.layer1(out)  # out (8,96,18,32) -> out (8,32,18,32)
        out = out.contiguous().view(x.size(0), -1)
        out = self.fc(out)
        return out  # out(8,8)
```

3 sorry, i don't understant what is 'homography in pixel space'.

4 i tried it again after adding  normalized_coordinates, and normalized_homography. 

images_warped = kgt.homography_warp(images.clone(), hg_mtxs, img_vt_s32_hg_shape, padding_mode="zeros", normalized_coordinates=False, normalized_homography=False) 

input  some images , after warpping most of the pixel values are 0, and a small part of the pixel values are not 0.
shows at below:
![image](https://github.com/kornia/kornia/assets/37241024/2dc5fdb4-36bd-478f-b274-10efcdaeabcb)

![image](https://github.com/kornia/kornia/assets/37241024/a660b44e-941d-420d-87ff-2c21e3bc30e5)

but other images, the result still black all pixel. for example:

![image](https://github.com/kornia/kornia/assets/37241024/83ee1ed0-8394-43b7-be91-d15c8a243024)

![image](https://github.com/kornia/kornia/assets/37241024/739b3483-86f1-4968-8e36-c464e1e344b2)

Thanks again for your help and reply.

**Reply by edgarriba** [reply]
1) In general Kornia assumes all the input images in range 0/1. But in this case as soon as your image is in floating point i believe it's fine, you will only have some issues with visualization. You might want to check `normalize_min_max` https://kornia.readthedocs.io/en/latest/enhance.html#kornia.enhance.normalize_min_max

2) Warp perspective transform a grid of point in pixel coordinates, for that reason the homography needs to reflect the mapping correspondence in pixel units. In your case what you can do is to denormalize the homography before applying the warping operator https://kornia.readthedocs.io/en/latest/geometry.conversions.html#kornia.geometry.conversions.denormalize_homography

As an alternative, you can use `warp_grid` https://github.com/kornia/kornia/blob/master/kornia/geometry/transform/imgwarp.py#L233
and `create_meshgrid` with `normalized_coordinates==False` https://kornia.readthedocs.io/en/latest/utils.html#kornia.utils.create_meshgrid

@ducha-aiki any other suggestions ? not sure if there are some properties of the homography you want to apply as regularization in the end of your network

**Reply by raphaelhou25** [reply]
thx, I’m truly grateful for your help. Your answer helped me solve a lot of problems, but, I still have some problems to discuss.

now, 

> firstly,  i garantee my homograpy matrix in [-1,1], 
> 
> secondly, use kornia.geometry.conversions.denormalize_homography  to reflect homograpy matrix in pixel units.
> 
> thirdly, i use kornia.geometry.transform.warp_perspective to warp image

**_but, i find in second step, after denormalize_homography , two elements in the matrix is so greater, I'm not sure whether this is a mistake._**

my code shows at below:

`mtxs = kgc.denormalize_homography(mtxs, (576, 1024), (576, 1024))`

> mtxs is : 
> 
> [-0.0722, -0.3154, -0.1566, -0.1254, -0.2406,  0.1205, -0.8795, 0.0946,  0.0112]
> 
> i got the result matrix is:
> 
> [-84.792114, -35.01265, _**46810.242**_, -50.326538, -13.011551, _**32855.86**_, -0.15319824, 0.029327393, 70.9375]

what i am confused about is that my image size is (576,1024),  but homograph matrix after 'denormalize_homography', 46810.242, 32855.86 , are so greater. whether it is a mistake. 

thx.

**Reply by edgarriba** [reply]
can you share a colab reproducing the example ?

**Reply by raphaelhou25** [reply]
@edgarriba  https://colab.research.google.com/drive/11cpleQD1AH669uEsR3s7XczdK_EnB6TS?usp=sharing

Please help me when you have time, 

Q1 I can't get the same result with opencv and with kornia.


Q2 and, i checked the apis of konia and opencv, i find a difference is Matrix M.

why 'kornia.geometry.transform.warp_perspective' is M^-1 and 'kornia.geometry.transform.warp_perspective3d' is M, and How is this different from opencv？

![image](https://github.com/kornia/kornia/assets/37241024/601dbc69-524e-40a9-9e8b-19cd13f3e562)
![image](https://github.com/kornia/kornia/assets/37241024/412747fc-eee0-4161-a722-5a63214811cb)
![image](https://github.com/kornia/kornia/assets/37241024/c1226023-7f27-4e9e-b6f9-5981be033013)

**Reply by edgarriba** [reply]
playing with your colab, seems that the matrix is already in pixel space. If you remove lines 15/16 in `kornia_wp` it works for me. FYi, to normalize the homography you just divide all the components by H_33
